crates/red_knot/src/lint.rs

```diff
@@ -140,6 +142,66 @@ fn lint_unresolved_imports(context: &SemanticLintContext) -> QueryResult<()> {
     Ok(())
 }
 
+fn lint_bad_overrides(context: &SemanticLintContext) -> QueryResult<()> {
+    // TODO possibly we should have a special marker on the real typing module (from typeshed) so
+    // if you have your own "typing" module in your project, we don't consider it THE typing module
+    let Some(typing_module) = context.db.resolve_module(ModuleName::new("typing"))? else {
+        debug_assert!(false, "typing module should always exist");
+        return Ok(());
+    };
+    let typing_file = context.db.module_to_file(typing_module)?;
+    let typing_table = context.db.symbol_table(typing_file)?;
+    let Some(typing_override) = typing_table.root_symbol_id_by_name("override") else {
+        debug_assert!(false, "typing.override should exist");
+        return Ok(());
+    };
+
+    // TODO we should maybe index definitions by type instead of iterating all, or else iterate all
+    // just once, match, and branch to all lint rules that care about a type of definition
+    for (symbol, definition) in context.symbols().all_definitions() {
+        if !matches!(definition, Definition::FunctionDef(_)) {
+            continue;
+        }
+        let ty = context
+            .db
+            .infer_definition_type(context.file_id, symbol, definition.clone())?;
+        let Type::Function(func) = ty else {
+            debug_assert!(false, "type of a FunctionDef should always be a Function");
+            continue;
```

Comment:
What's the motivation here for only panicking in a debug build? Ideally I feel like we wouldn't have to have an assertion like this, but if we do, I'd probably do something like this instead:

```suggestion
            unreachable!("type of a FunctionDef should always be a Function");
```

Reply:
Yeah, I wasn't sure about this, it's probably best to just use `unreachable`. Or even better to wrap this up in an API so most code doesn't have to make any assertion like this.